### PR TITLE
Add and remove ansible_role to host/hostgroup

### PR DESCRIPTION
**Pull request by Gauravtalreja1**
##### Description of changes

Adding add and remove ansible_role to host and hostgroup endpoints.

##### Upstream API documentation, plugin, or feature links
https://github.com/theforeman/foreman_ansible/pull/597
https://github.com/theforeman/foreman_ansible/pull/626

##### Additional Information

Any additional notes for reviewers, comments about the change, TODO lists on WIP PRs, etc.

### CODE REVIEW

File: nailgun/entities.py [hunk 1]
```diff
@@ -3702,6 +3708,89 @@ def rebuild_config(self, synchronous=True, timeout=None, **kwargs):
...
+        :param timeout: Maximum number of seconds to wait until timing out.
+            Defaults to ``nailgun.entity_mixins.TASK_TIMEOUT``.
+        :param kwargs: Arguments to pass to requests.
+        :returns: The server's response, with all JSON decoded.
+        :raises: ``requests.exceptions.HTTPError`` If the server responds with
+            an HTTP 4XX or 5XX message.
+        """
+        kwargs = kwargs.copy()
+        kwargs.update(self._server_config.get_client_kwargs())
+        #path = f'{self.path("ansible_roles")}/{kwargs["data"].pop("ansible_role_id")}'
```
Comment (Griffin-Sullivan) [inline nailgun/entities.py:3768]:
```suggestion
```

File: nailgun/entities.py [hunk 2]
```diff
@@ -4727,6 +4816,47 @@ def list_ansible_roles(self, synchronous=True, timeout=None, **kwargs):
...
```
Comment (Griffin-Sullivan) [inline nailgun/entities.py:4834]:
```suggestion
```

**Review by Griffin-Sullivan — APPROVED** [review]

File: nailgun/entities.py (hunk 1 above)
Comment (Gauravtalreja1) [reply nailgun/entities.py:3768]: The fix for this path is not yet in Snap, will uncomment it when the fix will be delivered to us

Comment (Griffin-Sullivan) [reply nailgun/entities.py:3768]: Oh ok, so is this not functional at the moment?

Comment (Gauravtalreja1) [reply nailgun/entities.py:3768]: the path below this line is functional and a commented path will be functional after the fix

Comment (Griffin-Sullivan) [reply nailgun/entities.py:3768]: OK would you like to merge as is and update when the snap is in?

**Review by synkd — APPROVED** [review]

File: nailgun/entities.py (hunk 1 above)
Comment (Gauravtalreja1) [reply nailgun/entities.py]: Updated, tested locally with snap 1 it is fixed, and in upcoming stream snap it should be there as well, so we're good to merge it now

**Review by Griffin-Sullivan — APPROVED** [review]

**Review by adarshdubey-star — DISMISSED** [review]

**Review by jyejare — CHANGES_REQUESTED** [review]
You haven't added enough coverage for all the functions you added !

File: tests/test_entities.py [hunk 3]
```diff
@@ -2899,33 +2905,64 @@ def test_read(self):
...
         * Method passes the right arguments to `client.*` and special
-            argument 'puppetclass_id' removed from data dict.
+            argument 'ansible_role_id' removed from data dict.
         * Method calls `entities._handle_response` once.
         * The result of `_handle_response(…)` is the return value.
+        """
+        entity = self.entity
+        entity.id = 1
+        func_param_dict = {entity.add_ansible_role: 'ansible_role_id'}
+        for func in list(func_param_dict.keys()):
```
Comment (omkarkhatavkar) [inline tests/test_entities.py:2923]: No need to cast to list, keys() returns the list object internally, you even can use the items() method and catch the key and value.

File: tests/test_entities.py [hunk 4]
```diff
@@ -2899,33 +2905,64 @@ def test_read(self):
...
-        self.assertEqual(len(client_request.call_args[0]), 1)
-        self.assertNotIn('puppetclass_id', client_request.call_args[1]['data'])
-        self.assertEqual(client_request.call_args[1], kwargs)
-        self.assertEqual(handlr.call_count, 1)
-        self.assertEqual(handlr.return_value, response)
+        func_param_dict = {
+            entity.delete_puppetclass: 'puppetclass_id',
+            entity.remove_ansible_role: 'ansible_role_id',
+        }
+        for func in list(func_param_dict.keys()):
```
Comment (omkarkhatavkar) [inline tests/test_entities.py:2954]: same here

**Review by omkarkhatavkar — APPROVED** [review]
ACK, small change needed

**Review by adarshdubey-star — APPROVED** [review]
Ack , Pending comment from Omkar.

**Review by omkarkhatavkar — APPROVED** [review]

**Comment by omkarkhatavkar** [comment]
@Gauravtalreja1 we need 1 more ack for merge

File: tests/test_entities.py (hunk 3 above)
Comment (Gauravtalreja1) [reply tests/test_entities.py]: updated

**Review by synkd — APPROVED** [review]

**Pull request merged**